experimental/virtio/src/queue.rs

```diff
@@ -0,0 +1,606 @@
+    idx: u16,
+    /// The ring-buffer containing the used elements.
+    ring: [UsedElem; QUEUE_SIZE],
+}
+
+impl<const QUEUE_SIZE: usize> Default for UsedRing<QUEUE_SIZE> {
+    fn default() -> Self {
+        Self {
+            flags: RingFlags::empty(),
+            idx: 0,
+            ring: [(); QUEUE_SIZE].map(|_| UsedElem::default()),
+        }
+    }
+}
+
+/// An element indicating a used descriptor chain.
+///
+/// See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-430008>.
+#[repr(C)]
+#[derive(Clone, Copy, Default)]
+struct UsedElem {
+    /// The index of the head of the used descriptor chain.
+    id: u32,
+    /// Total length of the bytes that was written to the used descriptor chain (only used for
+    /// device-write-only descriptors).
+    len: u32,
+}
+
+/// A split virtqueue implementation.
+///
+/// See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-240006>.
+#[repr(C, align(64))]
+struct VirtQueue<const QUEUE_SIZE: usize> {
+    /// The descriptor table.
+    desc: [Desc; QUEUE_SIZE],
+    /// The available ring, which is controlled by the driver.
+    avail: AvailRing<QUEUE_SIZE>,
+    /// The used ring, which is controlled by the device.
+    used: UsedRing<QUEUE_SIZE>,
+}
+
+/// The intended use for the queue.
+#[derive(PartialEq, Eq)]
+enum QueueType {
+    DeviceWriteOnly,
+    DriverWriteOnly,
+}
+
+/// A container for a boxed virtqueue and its associated buffers.
+pub struct Queue<const QUEUE_SIZE: usize, const BUFFER_SIZE: usize> {
+    /// The type of queue (whether the device or the driver can write to the buffers).
+    queue_type: QueueType,
+
+    /// The underlying virtqueue.
+    ///
+    /// We store it on the global heap for now, but in future we would have a dedicated allocator
+    /// to ensure it is in memory pages that are shared with the host.
+    virt_queue: Box<VirtQueue<QUEUE_SIZE>>,
+
+    /// The global buffer used by the virtqueue. Each descriptor uses a slice with an offset into
+    /// this single buffer for it own buffer.
+    ///
+    /// We store it as a `Vec` on the global heap for now, but in future we would have a dedicated
+    /// allocator to ensure it is located in memory pages that are shared with the host.
+    buffer: Vec<u8>,
+
+    /// The address of the first byte in the global buffer.
+    base_offset: u64,
+
+    /// The last index that was used when popping elements from the used ring.
+    last_used_idx: u16,
+
+    /// The indices of free descriptors that can be used by the driver for writing.
+    free_ids: VecDeque<u16>,
+}
+
+impl<const QUEUE_SIZE: usize, const BUFFER_SIZE: usize> Queue<QUEUE_SIZE, BUFFER_SIZE> {
+    /// Creates a new queue where the descriptor buffers are only writable by the driver.
+    ///
+    /// This is typically used for transmission queues.
+    pub fn new_driver_write_only() -> Self {
+        let mut result = Self::new(QueueType::DriverWriteOnly);
```

Comment:
I added to two wrapper types. It feels a bit clunky because of having to expose the inner queue as public to avoid a lot of code duplication, but at least the asserts are no longer required.